File: open3d_manage/Method/knn.py
import numpy as np
import open3d as o3d
from typing import Union

def toKNNIdxs(points: Union[np.ndarray, list], knn_num: int) -> np.ndarray:
    if isinstance(points, list):
        points = np.array(points)

    is_flatten = len(points.shape) == 1

    if is_flatten:
        points = points.reshape(-1, 3)

    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(points)

    pcd_tree = o3d.geometry.KDTreeFlann(pcd)

    knn_idxs = []
    for i in range(points.shape[0]):
        [_, knn_idx, _] = pcd_tree.search_knn_vector_3d(points[i], knn_num)
        knn_idxs.append(knn_idx)

    knn_idxs = np.array(knn_idxs, dtype=np.int64)

    if is_flatten:
        knn_idxs = knn_idxs.reshape(-1)

    return knn_idxs

def toKNNIdxsList(points: list, knn_num: int) -> list:
    return toKNNIdxs(points, knn_num).tolist()

File: src/knn.cpp
#include "knn.h"
#include "pcd.h"
#include <memory>
#include <open3d/Open3D.h>
#include <open3d/visualization/utility/DrawGeometry.h>

const nc::NdArray<int64_t> toKNNIdxs(const nc::NdArray<float> &points,
                                     const int &knn_num) {
  const int point_num = points.shape().rows;

  std::shared_ptr<open3d::geometry::PointCloud> pcd = toPcd(points);

  open3d::geometry::KDTreeFlann pcd_tree(*pcd);

  nc::NdArray<int64_t> knn_idxs = nc::zeros<int64_t>(point_num, knn_num);

  for (int i = 0; i < point_num; ++i) {
    std::vector<int> indices;
    std::vector<double> distances;

    pcd_tree.SearchKNN(pcd->points_[i], knn_num, indices, distances);

    for (int j = 0; j < indices.size(); ++j) {
      knn_idxs(i, j) = indices[j];
    }
  }

  return knn_idxs;
}
